src/hotspot/share/opto/library_call.cpp

```diff
@@ -4443,6 +4443,40 @@ void LibraryCallKit::arraycopy_move_allocation_here(AllocateArrayNode* alloc, No
     Node* alloc_mem = alloc->in(TypeFunc::Memory);
     C->gvn_replace_by(callprojs.fallthrough_ioproj, alloc->in(TypeFunc::I_O));
     C->gvn_replace_by(init->proj_out(TypeFunc::Memory), alloc_mem);
+
+    // Replace InitializeNode's output_control uses' control to AllocateArrayNode's input control.
+    // CastIINode might created in GraphKit::new_array (in AllocateArrayNode::make_ideal_length) and later uses.
+    // CastIINode's input_control should still use InitializeNode's output_control after move allocation.
+    // Otherwise it breaks constraint assumption.
+    //
+    // Because alloc is tightly coupled, checking InitializeNode's control proj's output for
+    // for CastIINode. After finding CastIINode, actions include:
+    // 1. Replace CastIINode with AllocateArrayNode's length.
+    // 2. Create CastIINode again in arraycopy_move_allocation_here after control flow adjustion.
+    Node* init_control = init->proj_out(TypeFunc::Control);
+    Node* alloc_length = alloc->Ideal_length();
+#ifdef ASSERT
+    Node* prev_cast = NULL;
+#endif
+    for (uint i = 0; i < init_control->outcnt(); i++) {
+      Node *init_out = init_control->raw_out(i);
+      if (init_out->is_CastII() && init_out->in(0) == init_control && init_out->in(1) == alloc_length) {
+#ifdef ASSERT
+        if (prev_cast == NULL) {
+          prev_cast = init_out;
+        } else {
+          // CastII must be same
+          CastIINode* prev = prev_cast->as_CastII();
+          CastIINode* cur = init_out->as_CastII();
+          assert(prev->has_range_check() == cur->has_range_check(), "not same");
+          assert(prev->type()->is_int()->_lo == cur->type()->is_int()->_lo, "not same");
+          assert(prev->type()->is_int()->_hi == cur->type()->is_int()->_hi, "not same");
+        }
```

Comment:
Is this really necessary? Have you seen cases with multiple identical CastII nodes? Or is it to be extra cautious?

Reply:
Thanks for your comment!

Multiple identical CastII node do happens and check is added to ensure they are identical CastIINodes. Cases are:
1. CastIINode created in GraphKit::new_array 
2. CastIINode might created before inline_arraycopy if GraphKit::load_array_length is invoked. If new array's length is loaded multiple times before System.arraycopy, there could be multiple CastIINode.

Reply:
simple example
```
        byte[] result = new byte[len];   // first CastII in GraphKit::new_array
        System.arraycopy(path, begin, result, 0, result.length);  // second CastII  when processing "result.length"
```

Reply:
Could you use hash_find() instead? hash_find(prev) == hash_find(cur) maybe?

Reply:
Done.